File: 14-TypeScript/Grey Multi-Tenant/packages/grey-dart/lib/src/domain/auth_client.dart
/// Auth domain client for Grey SDK.
library;

import 'dart:async';

import 'package:grpc/grpc.dart';

import '../error/grey_error.dart';
import '../grpc/auth_grpc_service.dart';
import '../grpc/grey_channel.dart';
import '../grpc/types.dart';

/// Client for authentication operations.
class AuthClient {
  /// Creates a new auth client.
  AuthClient(this._channel, this._service);

  final GreyChannel _channel;
  final AuthGrpcService _service;

  /// Authenticates a user with email and password.
  ///
  /// On success, stores the access token for subsequent authenticated requests.
  ///
  /// Throws [GreyError] on failure.
  Future<AuthData> login({
    required String email,
    required String password,
  }) async {
    // Validate inputs
    if (email.isEmpty) {
      throw GreyError(
        code: GreyError.validationError,
        message: 'Email is required',
      );
    }
    if (password.isEmpty) {
      throw GreyError(
        code: GreyError.validationError,
        message: 'Password is required',
      );
    }

    try {
      final result = await _service.login(
        email: email,
        password: password,
      );

      // Store the access token
      _channel.setAccessToken(result.accessToken);

      return result;
    } on GrpcError catch (e) {
      throw GreyError.fromGrpcError(e);
    } catch (e) {
      throw GreyError.fromException(e);
    }
  }

  /// Logs out the current user.
  ///
  /// Clears the stored access token.
  ///
  /// Throws [GreyError] on failure.
  Future<void> logout() async {
    try {
      await _service.logout();
    } on GrpcError catch (e) {
      throw GreyError.fromGrpcError(e);
    } catch (e) {
      throw GreyError.fromException(e);
    } finally {
      // Always clear the token
      _channel.setAccessToken(null);
    }
  }

  /// Refreshes the authentication tokens.
  ///
  /// If [refreshToken] is provided, uses it; otherwise uses the last known
  /// refresh token.
  ///
  /// Throws [GreyError] on failure.
  Future<AuthData> refresh({String? refreshToken}) async {
    final token = refreshToken;
    if (token == null || token.isEmpty) {
      throw GreyError(
        code: GreyError.validationError,
        message: 'Refresh token is required',
      );
    }

    try {
      final result = await _service.refresh(refreshToken: token);

      // Update the access token
      _channel.setAccessToken(result.accessToken);

      return result;
    } on GrpcError catch (e) {
      throw GreyError.fromGrpcError(e);
    } catch (e) {
      throw GreyError.fromException(e);
    }
  }

  /// Whether the client has a valid access token.
  bool get isAuthenticated => _channel.isAuthenticated;
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-scala/src/main/scala/com/grey/sdk/domain/AuthClient.scala
package com.grey.sdk.domain

import com.grey.sdk.error.{GreyError, GreyResult}
import com.grey.sdk.grpc.{AuthData, AuthGrpcService, GreyChannel}
import io.grpc.{StatusException, StatusRuntimeException}
import scala.concurrent.{ExecutionContext, Future}

/** Client for authentication operations.
  *
  * Provides login, logout, and token refresh functionality.
  */
final class AuthClient(
    channel: GreyChannel,
    service: AuthGrpcService
)(using ec: ExecutionContext):

  /** Authenticates a user with email and password.
    *
    * On success, stores the access token for subsequent authenticated requests.
    *
    * @param email
    *   The user's email address
    * @param password
    *   The user's password
    * @return
    *   Future containing AuthData on success
    * @throws GreyError
    *   on failure
    */
  def login(email: String, password: String): Future[AuthData] =
    // Validate inputs
    if email.isEmpty then
      return Future.failed(GreyError.validation("Email is required"))
    if password.isEmpty then
      return Future.failed(GreyError.validation("Password is required"))

    service.login(email, password).map { result =>
      channel.setAccessToken(Some(result.accessToken))
      result
    }.recoverWith { case e => Future.failed(GreyError.fromThrowable(e)) }

  /** Authenticates and returns Either instead of throwing.
    *
    * @return
    *   Future containing Either[GreyError, AuthData]
    */
  def loginSafe(email: String, password: String): Future[Either[GreyError, AuthData]] =
    login(email, password)
      .map(Right(_))
      .recover { case e: GreyError => Left(e) }

  /** Logs out the current user.
    *
    * Clears the stored access token.
    *
    * @return
    *   Future completing on success
    * @throws GreyError
    *   on failure
    */
  def logout(): Future[Unit] =
    service.logout()
      .map { _ => channel.setAccessToken(None) }
      .recoverWith { case e =>
        channel.setAccessToken(None) // Always clear token
        Future.failed(GreyError.fromThrowable(e))
      }

  /** Logs out and returns Either instead of throwing. */
  def logoutSafe(): Future[Either[GreyError, Unit]] =
    logout()
      .map(Right(_))
      .recover { case e: GreyError => Left(e) }

  /** Refreshes the authentication tokens.
    *
    * @param refreshToken
    *   Optional refresh token; uses last known if not provided
    * @return
    *   Future containing new AuthData on success
    * @throws GreyError
    *   on failure
    */
  def refresh(refreshToken: Option[String] = None): Future[AuthData] =
    refreshToken match
      case Some(token) if token.nonEmpty =>
        service.refresh(token).map { result =>
          channel.setAccessToken(Some(result.accessToken))
          result
        }.recoverWith { case e => Future.failed(GreyError.fromThrowable(e)) }
      case _ =>
        Future.failed(GreyError.validation("Refresh token is required"))

  /** Refreshes and returns Either instead of throwing. */
  def refreshSafe(refreshToken: Option[String] = None): Future[Either[GreyError, AuthData]] =
    refresh(refreshToken)
      .map(Right(_))
      .recover { case e: GreyError => Left(e) }

  /** Whether the client has a valid access token. */
  def isAuthenticated: Boolean = channel.isAuthenticated

object AuthClient:
  def apply(channel: GreyChannel, service: AuthGrpcService)(using ExecutionContext): AuthClient =
    new AuthClient(channel, service)
